### ISSUE TITLE
math/big: Implement fft algorithm

### ISSUE THREAD

**Issue by SparrowLii** [golang/go#42267]
From #30943. I think it's time for us to add fft to the big package. This implements Schönhage–Strassen algorithm in math/big and increased the multiplication speed several times to tens of times (depending on the length of the input). Although it is still not as fast as in GMP( which uses an advanced and more complex algorithm), it should solve the bottleneck of Go in multiplication of very large numbers(#30943, for example).
The benchmark test data is as follows:
Mul-4                 12.5ms ± 1%      6.9ms ± 3%  -45.08%  (p=0.016 n=4+5)
NatMul/10-4            251ns ± 3%      267ns ± 6%   +6.28%  (p=0.008 n=5+5)
NatMul/100-4          8.64µs ± 3%     8.80µs ± 1%     ~     (p=0.310 n=5+5)
NatMul/1000-4          358µs ± 5%      355µs ± 3%     ~     (p=0.841 n=5+5)
NatMul/10000-4        12.6ms ± 2%      7.0ms ± 5%  -44.11%  (p=0.008 n=5+5)
NatMul/100000-4        545ms ± 4%       90ms ± 3%  -83.43%  (p=0.008 n=5+5)
NatMul/1000000-4       21.2s ± 2%       1.2s ± 4%  -94.33%  (p=0.008 n=5+5)

**Comment by SparrowLii** [comment]
@bmkessler Hi brain, could you help looking it? @griesemer

**Comment by advancedwebdeveloper** [comment]
@dcmartin , could give an advice here?

**Comment by client9** [comment]
Thank you @SparrowLii and other reviewers.

I took a look at this.  I'm not setup to send patches in, yet, but until then:

* I ripped out math/big into separate repo for hacking: https://github.com/client9/big
* Modified patch set 4 from  https://golang.org/cl/266201 to work with the latest go 1.25.1 source
* It's in `natmul_fft.go`  https://github.com/client9/big/blob/main/natmul_fft.go
* Patch works with only minor changes:

```
// Needed to add back this function.  I assume to be replaced with native 'clear'
func (z nat) clear() {
        for i := range z {
                z[i] = 0
        }
}

// Placeholder.  Need to review how migration of shlVU to lshVU was done
//
// shlVU is obsolete, but existing lshVU doesn't handle s==0 case.
// also  shlVU also isn't ported to a pure go implementation so
//     go test -tags math_big_pure_go
// will fail.  This local function works until correct replacement.
//
//
func local_shlVU(z, x []Word, s uint) (c Word) {
        if s == 0 {
                copy(z, x)
                return 0
        }
        return lshVU(z, x, s)
}
```

Also multiplication did not use the nat stack (the cache/pool), so modified by passing in`nil` for now

```
t := nat(nil).mul(nil, a[:nprime], b[:nprime])
```

Otherwise, works, with performance improvements starting around 2500 "words" (160k bits, 50k decimal digits) on my Apple M4.  Memory usage is not so great, as expected.

Once I get the:

- [x]  use nat `stack` correctly
- [x]  use native `clear` 
- [x]  figure out correct `lshVU` replacement
- [x]  add a fuzz test for correctness and coverage, comparing existing vs. fft results.  Fuzzer will set a seed for a PRNG to generate random vectors to test.

then I'll repost.

After that, we can review the recommendations by Brian Kessler and Pavel Kositsyn.

Onwards!

nick g


```
goos: darwin
goarch: arm64
pkg: github.com/client9/big
cpu: Apple M4
                  │   old.txt    │               new.txt               │
                  │    sec/op    │   sec/op     vs base                │
NatMul/10-10         69.59n ± 0%   69.37n ± 0%   -0.33% (p=0.000 n=10)
NatMul/100-10        2.536µ ± 0%   2.540µ ± 2%        ~ (p=0.136 n=10)
NatMul/1000-10       104.3µ ± 0%   169.6µ ± 0%  +62.61% (p=0.000 n=10)
NatMul/2500-10       484.9µ ± 1%   406.1µ ± 0%  -16.25% (p=0.000 n=10)
NatMul/5000-10       1.446m ± 0%   1.111m ± 0%  -23.13% (p=0.000 n=10)
NatMul/10000-10      4.349m ± 2%   1.734m ± 0%  -60.12% (p=0.000 n=10)
NatMul/100000-10    153.93m ± 4%   24.21m ± 1%  -84.27% (p=0.000 n=10)
NatMul/1000000-10   6396.8m ± 0%   352.2m ± 0%  -94.49% (p=0.000 n=10)
geomean              696.3µ        344.7µ       -50.49%

                  │   old.txt    │                  new.txt                   │
                  │     B/op     │      B/op       vs base                    │
NatMul/10-10          192.0 ± 0%       192.0 ± 0%          ~ (p=1.000 n=10) ¹
NatMul/100-10       1.751Ki ± 0%     1.751Ki ± 0%          ~ (p=0.211 n=10)
NatMul/1000-10      16.01Ki ± 0%    235.34Ki ± 0%  +1370.04% (p=0.000 n=10)
NatMul/2500-10      40.04Ki ± 0%    539.17Ki ± 0%  +1246.57% (p=0.000 n=10)
NatMul/5000-10      80.29Ki ± 0%   1358.09Ki ± 0%  +1591.58% (p=0.000 n=10)
NatMul/10000-10     161.5Ki ± 1%    2095.8Ki ± 0%  +1197.75% (p=0.000 n=10)
NatMul/100000-10    2.299Mi ± 0%    20.406Mi ± 0%   +787.77% (p=0.000 n=10)
NatMul/1000000-10   68.39Mi ± 0%    169.25Mi ± 0%   +147.46% (p=0.000 n=10)
geomean             67.86Ki          379.4Ki        +459.09%
¹ all samples are equal

                  │   old.txt   │                   new.txt                   │
                  │  allocs/op  │   allocs/op    vs base                      │
NatMul/10-10        1.000 ±  0%      1.000 ± 0%            ~ (p=1.000 n=10) ¹
NatMul/100-10       1.000 ±  0%      1.000 ± 0%            ~ (p=1.000 n=10) ¹
NatMul/1000-10      1.000 ±  0%    273.000 ± 0%   +27200.00% (p=0.000 n=10)
NatMul/2500-10      1.000 ±  0%    530.000 ± 0%   +52900.00% (p=0.000 n=10)
NatMul/5000-10      1.000 ±  0%   1043.000 ± 0%  +104200.00% (p=0.000 n=10)
NatMul/10000-10     1.000 ±  0%   1045.000 ± 0%  +104400.00% (p=0.000 n=10)
NatMul/100000-10    2.000 ±  0%   4123.000 ± 0%  +206050.00% (p=0.000 n=10)
NatMul/1000000-10   10.00 ± 10%    8220.00 ± 0%   +82100.00% (p=0.000 n=10)
geomean             1.454            219.3        +14978.63%
```

**Comment by client9** [comment]
Ok!  

- recycling a vector dropped memory by 45%.
- using the `big.stack` made nice improvements in both memory and performance for vectors under 10,000 words.  After that it does nothing.  The  stack >= 64Kb is maybe sent to the heap, or `sync.Pool` ejects them.   To be investigated later.

Major Issues:
- [ ] Logic issue in setup.  Values of N and K are dependent on the `fftThreshold`, but that's just an arbitrary value for performance tuning.  N,K should be chosen based on input size.  This section needs more comments/documentation as well.

Minor Issues:
- [ ]  Initial set up creates some static arrays based on N,K.  These could be cached and reused.  The setup should be pulled out into a separate function anyways.
- [x] There are a few unnecessary slice clearings or copies
- [x]  use range instead of c-style for loops
- [x] Use `max()` instead of `if/else` in a few places
- [x] Nesting simplification.  Lots of `if ... { block1 } else { block2 }`  (C style) which could be converted to `if ... { block1; return } block2` (Go style)
- [x] figure out correct `lshVU` replacement -- currently works fine but will probably needs some adjustment
- [ ] code comments showing math/algorithm got messed up with tabs vs spaces.  Cleanup, use superscripts, etc.

After this, we can work on a memory allocation.  The code still has a lot of allocation:

```go
        A := nat(nil).make(K * (nprime + 1)) // storage for fft
        B := nat(nil).make(K * (nprime + 1))
        Ap := make([]nat, K)
        Bp := make([]nat, K)
        T := nat(nil).make(2*nprime + 2) // temporary storage
```

Latest:
```
                 │   old.txt    │               new3.txt               │
                  │    sec/op    │    sec/op     vs base                │
NatMul/10-10         69.59n ± 0%   70.17n ±  1%   +0.83% (p=0.002 n=10)
NatMul/100-10        2.536µ ± 0%   2.542µ ±  1%        ~ (p=0.362 n=10)
NatMul/1000-10       104.3µ ± 0%   104.9µ ±  0%   +0.57% (p=0.000 n=10)
NatMul/2500-10       484.9µ ± 1%   357.8µ ±  0%  -26.20% (p=0.000 n=10)
NatMul/5000-10      1445.7µ ± 0%   767.4µ ±  0%  -46.91% (p=0.000 n=10)
NatMul/10000-10      4.349m ± 2%   1.649m ±  0%  -62.08% (p=0.000 n=10)
NatMul/100000-10    153.93m ± 4%   23.42m ± 12%  -84.78% (p=0.000 n=10)
NatMul/1000000-10   6396.8m ± 0%   348.7m ±  0%  -94.55% (p=0.000 n=10)
geomean              696.3µ        302.0µ        -56.63%

                  │   old.txt    │                 new3.txt                 │
                  │     B/op     │     B/op       vs base                   │
NatMul/10-10          192.0 ± 0%      192.0 ± 0%         ~ (p=1.000 n=10) ¹
NatMul/100-10       1.751Ki ± 0%    1.751Ki ± 0%         ~ (p=1.000 n=10) ¹
NatMul/1000-10      16.01Ki ± 0%    16.01Ki ± 0%         ~ (p=1.000 n=10)
NatMul/2500-10      40.04Ki ± 0%   298.60Ki ± 0%  +645.75% (p=0.000 n=10)
NatMul/5000-10      80.29Ki ± 0%   604.76Ki ± 0%  +653.26% (p=0.000 n=10)
NatMul/10000-10     161.5Ki ± 1%   1200.0Ki ± 0%  +643.04% (p=0.000 n=10)
NatMul/100000-10    2.299Mi ± 0%   11.405Mi ± 0%  +396.16% (p=0.000 n=10)
NatMul/1000000-10   68.39Mi ± 0%    95.25Mi ± 0%   +39.27% (p=0.000 n=10)
geomean             67.86Ki         183.7Ki       +170.71%
¹ all samples are equal

                  │   old.txt   │                new3.txt                 │
                  │  allocs/op  │  allocs/op   vs base                    │
NatMul/10-10        1.000 ±  0%    1.000 ± 0%          ~ (p=1.000 n=10) ¹
NatMul/100-10       1.000 ±  0%    1.000 ± 0%          ~ (p=1.000 n=10) ¹
NatMul/1000-10      1.000 ±  0%    1.000 ± 0%          ~ (p=1.000 n=10) ¹
NatMul/2500-10      1.000 ±  0%   18.000 ± 0%  +1700.00% (p=0.000 n=10)
NatMul/5000-10      1.000 ±  0%   19.000 ± 0%  +1800.00% (p=0.000 n=10)
NatMul/10000-10     1.000 ±  0%   21.000 ± 0%  +2000.00% (p=0.000 n=10)
NatMul/100000-10    2.000 ±  0%   24.000 ± 0%  +1100.00% (p=0.000 n=10)
NatMul/1000000-10   10.00 ± 10%    28.00 ± 4%   +180.00% (p=0.000 n=10)
geomean             1.454          6.846        +370.79%
```

**Comment by chadsec1** [comment]
Hi, sorry for disturbing the flow of discussion, but, may I ask when will the Schönhage–Strassen fft implementation be available ?

**Comment by client9** [comment]
My private version works @  https://github.com/client9/big  
Final issue is really making is finalizing parameters for performance tuning.  
I stepped away from it over the last month due to other priorities but I'll get back to it in 2026.

**Comment by chadsec1** [comment]
> My private version works @ https://github.com/client9/big Final issue is really making is finalizing parameters for performance tuning. I stepped away from it over the last month due to other priorities but I'll get back to it in 2026.

Ah thanks. I was asking when it would be available for everyone in math/big it's self.
Thanks for all your work, really.

**Comment by griesemer** [comment]
@chadsec1 Somebody from the Go team will need to review/vet this and we don't really have the bandwidth for this at the moment. It's not obvious that this is the most pressing Go library issue for the vast majority of Go users, so it will have to wait. Since there's a private version (see above), how about using that one?

**Comment by chadsec1** [comment]
> [@chadsec1](https://github.com/chadsec1) Somebody from the Go team will need to review/vet this and we don't really have the bandwidth for this at the moment. It's not obvious that this is the most pressing Go library issue for the vast majority of Go users, so it will have to wait. Since there's a private version (see above), how about using that one?

Well, I much rather see it merged in Go std lib, because then, I don't need to worry as much about my dependencies (in terms of security).

(It is pressing incase someone want to implement a database implementation in Go, they'd have to use a third party repo for something essentinal like this., also it might improve a lot of Go programs already excellent peformance)

**Comment by griesemer** [comment]
> It is pressing incase someone want to implement a database implementation in Go, they'd have to use a third party repo for something essentinal like this

Can you explain why this is _essential_ for a DB? My understanding is that the benefit only shows up for very large numbers. I am not trying to be facetious here, it's meant as a serious question. I probably don't understand the use case. Thanks.

**Comment by chadsec1** [comment]
> > It is pressing incase someone want to implement a database implementation in Go, they'd have to use a third party repo for something essentinal like this
> 
> Can you explain why this is _essential_ for a DB? My understanding is that the benefit only shows up for very large numbers. I am not trying to be facetious here, it's meant as a serious question. I probably don't understand the use case. Thanks.

https://gitlab.com/cznic/sqlite/-/issues/229

**Comment by griesemer** [comment]
Thanks, but [this](https://gitlab.com/cznic/sqlite/-/issues/229) is a specific issue for a specific DB. The DB will work fine even w/o fast multiplication; it's an optimization at best. It would be nice it if were faster, but that was not what I was asking for.

I am trying to understand if there is a _broad_ need (not just by you) of users of this DB, but more importantly, general users of math/big, for very large integer multiplication. I understand that _some_ people would like it (including myself, if just for the fun of it), but that is not the same as this is a critical issue for the Go std library.

Again, just to be clear, we are not against this, but we do not have limitless resources/people in the Go team and hence this is simply not a high priority.

**Comment by chadsec1** [comment]
> Thanks, but [this](https://gitlab.com/cznic/sqlite/-/issues/229) is a specific issue for a specific DB. The DB will work fine even w/o fast multiplication; it's an optimization at best. It would be nice it if were faster, but that was not what I was asking for.
> 
> I am trying to understand if there is a _broad_ need (not just by you) of users of this DB, but more importantly, general users of math/big, for very large integer multiplication. I understand that _some_ people would like it (including myself, if just for the fun of it), but that is not the same as this is a critical issue for the Go std library.
> 
> Again, just to be clear, we are not against this, but we do not have limitless resources/people in the Go team and hence this is simply not a high priority.

You're right, it's not essential, but it kinda is. Would you use a slow database? 
Well, let's ignore databases for now, what about scientific applications that deal with large numbers? etc
Nothing is truly essential with computers, we could argue about a lot of things that are "essential" but in reality just optimizations at best. 
Take Go for instance, it's pretty cool language, probably my favourite language, from my perspective, go is plenty complete, and should go into maintenance mode for stdlib etc. Now that previous take is pretty bad, no ?

It has many uses. None are really essential in the sense, a lot of third party exists, but it would be excellent if Go were to implement this in stdlib because it is pretty widely used in a lot of applications, and ppl turn to third party, which might have questionable implementations, which may or may not introduce vulnerabilities, and also questionable supply chain, which may or may not introduce subtle backdoors. It's sounds far fetched, but it's true, and it will only get more common.. That's why I raised that issue on that db, security. 

TL;DR: It's widely used + appreicated, but not essential per say because there exists third party libraries, but most of them are proof of concepts, and more importantly, questionable in terms of security because there are no eyes on them.

I didn't know Go has limited resources/people. Any way I can get involved? I mean, putting pull requests would be almost useless as demonstrated here because you guys have to spend time to review, validate, etc. I meant, get involved as, get maybe hired ?
Sorry for backhanded job searching offer,... it's tough out here in the market haha

**Comment by griesemer** [comment]
@chadsec1 I am not disagreeing with you. But the Go team is surprisingly small and lean and we have a lot of things that we need to do and a lot of things that we want to do. We can't do all of it, not even close. So we need to prioritize, and FFT for multiplication is unfortunately low on the priority list. Thanks.

The best way to get involved is via open-source contributions: bug fixes, etc. For hiring, please use the regular application channels.
